Support thread-safety with ARMC6

ccli8 opened the pull request:
### Description

This PR tries to add thread-safety support with **ARMC6** toolchain.

#### Related issue
https://github.com/ARMmbed/mbed-os/issues/6396
https://github.com/ARMmbed/mbed-cloud-client-example/issues/11


### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

geky commented:
~This looks good to me, but not really my area of expertise~ :+1: 

Ah! Sorry it was my first time seeing a review request for a group, glad to see that's working now, should help quite a bit with reviews moving forward

kegilbert commented on rtos/TARGET_CORTEX/mbed_boot.c:
```diff
@@ -507,6 +507,41 @@ void _mutex_free(mutex *m) {
+#if (defined(__ARMCC_VERSION) && (__ARMCC_VERSION >= 6010050))
+
+// Check if Kernel has been started
+static uint32_t os_kernel_is_active (void) {
```
Minor nit: Spacing in this function is inconsistent with the rest of the file (2 spaces versus 4).

ccli8 replied:
@kegilbert I fix spacing in https://github.com/ARMmbed/mbed-os/pull/6973/commits/a775f1830e1f22f0f42b2c9fe3fba06b2c6255b9.

bulislaw commented on rtos/TARGET_CORTEX/mbed_boot.c:
```diff
@@ -507,6 +507,41 @@ void _mutex_free(mutex *m) {
+// Acquire mutex
+__USED void _mutex_acquire(mutex *m)
+{
+    if (os_kernel_is_active()) {
```
Why do we need to check that?

ccli8 replied:
@bulislaw This code is copied from `mbed-os/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c`. Maybe its original owner could comment it well. I guess it is to guard from e.g. C/C++ initialization code during which kernel is not ready yet.

bulislaw reviewed (commented):
Actually why do we need to redefine it? The same functions from rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c should be used by default

geky commented:
From https://github.com/ARMmbed/mbed-os/issues/6396#issuecomment-391465728, I think the correct solution is to make the versions in rtx_lib.c available on all versions of ARMCC:
https://github.com/ARMmbed/mbed-os/blob/master/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c#L623

ccli8 commented:
As commented, I force `_mutex_acquire`/`_mutex_release` to `__USED` for **ARM**/**ARMC6** toolchains in `rtx_lib.c` and restore `mbed_boot.c` (no `_mutex_acquire`/`_mutex_release` are re-defined here). But I don't also force `_mutex_initialize`/`_mutex_free` to `__USED` because I meet error with **ARMC6** toolchain:
```
Error: L6200E: Symbol _mutex_initialize multiply defined (by rtx_lib.o and mbed_boot.o).
Error: L6200E: Symbol _mutex_free multiply defined (by rtx_lib.o and mbed_boot.o).
```

It seems that with **ARMC6** toolchain the `__WEAK` modifier will be cancelled by the `__USED` modifier.

bulislaw commented:
@ccli8 we've just merged update of RTX could you retest the vanilla version and if it still doesn't work rebase your PR.

geky commented:
@ccli8, were the `__USED` and `__WEAK` on different declarations?

``` cpp
__USED _mutex_free(void*);
__WEAK _mutex_free(void *m) {
    Blah blah blah
}
```

deepikabhavnani commented:
> It seems that with ARMC6 toolchain the __WEAK modifier will be cancelled by the __USED modifier.

Can we try __WEAK in "Source/rtx_lib.c" definition and __USED in "mbed_boot.c" ?

ccli8 commented:
@bulislaw Where's vanilla version?

ccli8 commented:
> were the `__USED` and `__WEAK` on different declarations?

@geky I've tried it and in the same line. Still meet the **multiple definitions** link error with **ARMC6** toolchain.

ccli8 commented:
> Can we try __WEAK in "Source/rtx_lib.c" definition and __USED in "mbed_boot.c" ?

@deepikabhavnani It seems that doesn't make sense. `_mutex_`xxx functions defined in `rtx_lib.c` are to be re-implemented in `mbed_boot.c`, so they are qualified with `__WEAK`. Because **ARM**/**ARMC6** libraries refer **weakly** to them, they must be `__USED` to be present in the final image. So `_mutex_`xxx functions defined in `rtx_lib.c` must be qualified with `__WEAK` and `__USED` simultaneously. If we just have `__WEAK` in `rtx_lib.c` but the function is not re-implemented in `mbed_boot.c`, it won't linked in.

0xc0170 commented:
@ccli8 The above recommendation was to rebase this one (there is a conflict anyway) to get the latest rtx update , and try it with that one and report back.

ccli8 commented:
I do rebase to fix the conflict (https://github.com/ARMmbed/mbed-os/pull/6973/commits/80038346f34dfc9d856c42c14d3b054e0f307726). Just add `__USED` to `_mutex_acquire`/`_mutex_release` in `rtx_lib.c` for **ARMC6** toolchain and have no other change.

ccli8 commented:
Any update?

0xc0170 commented on rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c:
```diff
@@ -693,21 +693,15 @@ __WEAK int _mutex_initialize(mutex *m) {
-__USED
-#endif
-__WEAK void _mutex_acquire(mutex *m);
+__WEAK __USED void _mutex_acquire(mutex *m);
```
used was added to fix this issue https://github.com/ARM-software/CMSIS_5/issues/129. Looking at this, do we still need weak? (the PR referenced at the bottom there, was closed, and `RTX_NO_MULTITHREAD_CLIB ` was mentioned there).

JonatanAntoni replied:
__WEAK and __USED seems not to make sense if used both. As discussed earlier this function cannot be defined __WEAK because there is already a weak version provided in the C library. Hence the linker cannot distinguish which weak version to use if no non-weak version is given. If its really needed for mbed to get rid of the internal implementation that is part of RTX we need to find a completely different solution.

0xc0170 commented:
@JonatanAntoni Can you review please? We already looked at these a while ago, seems like there is a problem with ARMC6 now.

cmonr commented:
@JonatanAntoni Could you take another look at this?

JonatanAntoni commented:
What exactly should I take a look to?


Von meinem Samsung Galaxy Smartphone gesendet.


-------- Ursprüngliche Nachricht --------
Von: Cruz Monrreal <notifications@github.com>
Datum: 14.06.18 17:33 (GMT+01:00)
An: ARMmbed/mbed-os <mbed-os@noreply.github.com>
Cc: Jonatan Antoni <Jonatan.Antoni@arm.com>, Mention <mention@noreply.github.com>
Betreff: Re: [ARMmbed/mbed-os] Support thread-safety with ARMC6 (#6973)


@JonatanAntoni<https://github.com/JonatanAntoni> Could you take another look at this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ARMmbed/mbed-os/pull/6973#issuecomment-397338417>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AYmc6GZtv52MKUcUJJbFCp9MyH5vc2F4ks5t8oI2gaJpZM4UH73d>.

IMPORTANT NOTICE: The contents of this email and any attachments are confidential and may also be privileged. If you are not the intended recipient, please notify the sender immediately and do not disclose the contents to any other person, use it for any purpose, or store or copy the information in any medium. Thank you.

0xc0170 commented:
We need to look at this and find a solution. Using used and weak does not look correct as mentioned above.

@ARMmbed/mbed-os-core Can you review this PR ?

> What exactly should I take a look to?

@cmonr  missed the review comment above

JonatanAntoni commented:
Finally we sorted out the misunderstandings. We discussed the library integration last year, see https://github.com/ARM-software/CMSIS_5/pull/213. The conclusion was adding the macro `RTX_NO_MULTITHREAD_CLIB`. If you compile upstream version of RTX5 with this macro defined the default clib integration is skipped. Instead you can provide an mbed specific one. No changes to `rtx_lib.c` should be needed at all.

0xc0170 commented:
My proposal is that we remove `weak` addition for these 2 functions that are changed (we leave the same as in RTX upstream repo). We do not overwrite them in our codebase thus there is no strong symbol and linker complains. That should fix it. 
@ccli8 can you test it?

We should look at `RTX_NO_MULTITHREAD_CLIB` option as @JonatanAntoni proposing (no change in RTX as we do now for mutex initialize/free). Thus even this patch goes in, #6396 issue might stay open (or create a new issue about used + weak for _mutex_initialize/_mutex_free)

ccli8 commented:
> My proposal is that we remove weak addition for these 2 functions that are changed (we leave the same as in RTX upstream repo). We do not overwrite them in our codebase thus there is no strong symbol and linker complains. That should fix it.

@0xc0170 Remove `__WEAK` from `_mutex_acquire`/`_mutex_release` (in `rtx_lib.c`) is fine, as long as you are not to override them (in `mbed_boot.c`). But `__USED` is always necessary to avoid excluded by linker.

Besides, I find there will be similar issue with `_mutex_free`. With `_mutex_initialize`/`_mutex_free` being to override in `mbed_boot.c` and `_mutex_acquire`/`_mutex_release` being not, I would propose:

**rtx_lib.c:**
`__WEAK int _mutex_initialize(mutex *m);`
`__USED void _mutex_acquire(mutex *m);`
`__USED void _mutex_release(mutex *m);`
`__WEAK void _mutex_free(mutex *m);`

**mbed_boot.c:**
`__USED int _mutex_initialize(mutex *m);`
`__USED void _mutex_free(mutex *m);`

JonatanAntoni commented:
@ccli8, please be aware that your solution might work for mbed but we won't be able to upstream this change to official RTX. In official RTX we introduced the preprocessor define `RTX_NO_MULTITHREAD_CLIB` to achieve exactly what is needed here. Hence I strongly recommend to go this way and avoid diverging from upstream RTX.

ccli8 commented:
@JonatanAntoni Would you comment more with `RTX_NO_MULTITHREAD_CLIB`? How does it guarantee that `_mutex_acquire`/`_mutex_release` are `USED` with **ARMC6**?

JonatanAntoni commented:
The macro `RTX_NO_MULTITHREAD_CLIB` can be used to exclude the library retarget implementation shipped with the RTX5 source code. Instead you can provide your own implementation. You can refer to the RTX5 implementation, duplicate the code into mbed and adopt the functions you need to adopt.

The drawback is that you need to implement *all* functions on your own. You cannot selectively choose functions you want to change. This is the only viable way to keep application specific requirements out of RTX5 code base.

ccli8 commented:
@JonatanAntoni So to support thread-safety with **ARMC6**, I need to define `RTX_NO_MULTITHREAD_CLIB` and override all `_mutex_xxx` functions as `_mutex_initialize`/`_mutex_free` are overridden now?

JonatanAntoni commented:
@ccli8, yes. Just copy the implementation starting at [rtx_lib.c:610](https://github.com/ARMmbed/mbed-os/blob/94a06cef661105f2a5779690d4a2db70b7087678/rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_lib.c#L610)

bulislaw commented:
Can we make sure that the implementation is shared between ARMC5 and ARMC6, I don't think there are differences in the API.

JonatanAntoni commented:
@bulislaw as you can see in the RTX5 implementation it applies to all versions of Arm Compiler, i.e. 5 and 6. The only exception is using Microlib which has no multi thread interface at all.

ccli8 committed:
Support thread-safety with ARMC6

1. Define RTX_NO_MULTITHREAD_CLIB to provide Mbed-specific multi-thread support for ARM/ARMC6
2. All overridden _mutex_xxx functions are declared with __USED to avoid excluded by linker

NOTE: Microlib doesn't support multi-thread

ccli8 commented:
@JonatanAntoni https://github.com/ARMmbed/mbed-os/pull/6973/commits/afc0a28cc50223436b08f18a29b64f08b9154f8a is the re-implementation by following the `RTX_NO_MULTITHREAD_CLIB` way. It has the following modifications:
1. Define `RTX_NO_MULTITHREAD_CLIB` in `RTX_Config.h`, but I am not sure if it is the right place to add this macro.
2. Move all `RTX_NO_MULTITHREAD_CLIB` code from `rtx_lib.c` to `mbed_boot.c` except `_mutex_initialize`/`_mutex_free` which have been overridden there.
3. All `_mutex_xxx` functions are defined with `__USED` to avoid excluded by linker.

JonatanAntoni commented:
@ccli8, the commit is exactly what I had in mind. That should work. Its okay to put this configuration to `RTX_Config.h`.

0xc0170 commented:
@ccli8  👍  Thanks for resolving this issue !

/morph build

bulislaw commented on rtos/TARGET_CORTEX/rtx5/RTX/Config/RTX_Config.h:
```diff
@@ -404,6 +404,11 @@
 #endif
  
+
+// Don't adopt default multi-thread support for ARM/ARMC6 toolchains from RTX code base.
```
We are trying not to change the default config file. Could it be moved to rtos/TARGET_CORTEX/mbed_rtx_conf.h

ccli8 replied:
@bulislaw OK. Please see https://github.com/ARMmbed/mbed-os/pull/6973/commits/807140996392dcd61eeed2029eccc0054922059f.

bulislaw commented on rtos/TARGET_CORTEX/mbed_boot.c:
```diff
@@ -423,6 +423,60 @@ void __rt_entry (void) {
     mbed_start_main();
 }
 
+#if defined(RTX_NO_MULTITHREAD_CLIB)
```
Could you add a comment saying it was copied from whichever file in rtx.

ccli8 replied:
@bulislaw I add comment in https://github.com/ARMmbed/mbed-os/pull/6973/commits/f055c2a7006913215c509e740814ad4859268e05.

ccli8 committed:
Add comment with RTX_NO_MULTITHREAD_CLIB movement

ccli8 committed:
Move RTX_NO_MULTITHREAD_CLIB from RTX_Config.h to mbed_rtx_conf.h

This is to avoid change to RTX code base.

cmonr commented:
@bulislaw Would you mind re-reviewing?

bulislaw reviewed (approved):
Looks awesome!

deepikabhavnani reviewed (approved):
Looks good to me 👍

cmonr commented:
/morph build

pan- commented:
this patch cause an link error when an application is compiled with `MBED_RTOS_SINGLE_THREAD`.

pan- commented:
Seems like a root cause was due to a custom profile without `_RTE_` defined; there's nothing to see here.
